File: demos/android-demo/android/app/src/main/java/com/example/androidapp/act/ManagedRenderer.kt
package com.example.androidapp.act

import android.view.View
import com.example.androidapp.act.serializable.Commit
import com.example.androidapp.act.serializable.Diff
import com.example.androidapp.act.serializable.ElementComponent
import com.example.androidapp.act.serializable.FunctionComponent

class ManagedRenderer<Node>(var options: Options<Node>) {
    val nodesByCommitID = mutableMapOf<String, Node>()

    class Options<Node>(
        val add: (
            diff: Diff,
            element: ElementComponent
        ) -> Node?,
        val update: (
            diff: Diff,
            node: Node,
            children: List<RenderResult<Node>>
        ) -> Unit,
        val delete: (
            diff: Diff,
            node: Node,
            children: List<RenderResult<Node>>
        ) -> Unit,
    );
    class RenderResult<Node>(val node: Node, val commit: Commit);

    fun createNode(diff: Diff): Node? {
        val node = when (val component = diff.next.element.component) {
            is FunctionComponent -> null;
            is ElementComponent ->  options.add(diff, component);
            else -> null;
        }
        if (node != null)
            nodesByCommitID[diff.next.id] = node;
        return node
    };

    fun updateNode(diff: Diff, node: Node, children: List<RenderResult<Node>>) {
        options.update(diff, node, children);
    }


    fun removeNode(diff: Diff, node: Node, children: List<RenderResult<Node>>) {
        options.delete(diff, node, children);
        nodesByCommitID.remove(diff.next.id);
    }

    fun getNodes (commit: Commit): List<RenderResult<Node>> {
        var node = nodesByCommitID[commit.id];
        if (node != null)
            return listOf(RenderResult(node, commit));

        return commit.children.map(fun (c) = getNodes(c)).flatten();
    }

    fun render(diff: Diff, parentNode: Node? = null): List<RenderResult<Node>> {
        var node = nodesByCommitID[diff.next.id] ?: createNode(diff);
        var children = diff.diffs.map(fun (diff) = render(diff, node)).flatten();

        var hasDiff = diff.next.id != diff.prev.id;

        if (node == null) {
            return if (!hasDiff)
                getNodes(diff.next);
            else
                children;
        }

        updateNode(diff, node, children);

        if (!diff.next.pruned)
            return listOf(RenderResult(node, diff.next));

        removeNode(diff, node, children);
        return emptyList();
    }
}

File: demos/ios-demo/ios/act/ManagedRenderer.swift
//
//  ManagedRenderer.swift
//  ios
//
//  Created by Luke Kaalim on 12/12/2022.
//

import Foundation
import SwiftUI

public struct RenderResult {
    let commit: Commit;
    let node: UIView;
}

class ViewRenderer {
    
}


class ManagedRenderer {
    var nodes: Dictionary<String, UIView> = Dictionary();
    
    func createNode(_ diff: Diff) -> UIView? {
        let type = diff.next.element.component;
        switch (type) {
        case .function:
            return nil;
        case .element(let element):
            switch (element.name) {
            case "ios:text":
                return UILabel();
            case "ios:stack_view":
                return UIStackView();
            default:
                return nil;
            }
        }
    }
    
    func updateNode(node: UIView?, diff: Diff, children: [RenderResult]) {
        switch (node) {
        case let label as UILabel:
            let contentProp = diff.next.element.props["content"];
            if case .mixed(let prop) = contentProp! {
                label.text = prop.value.string!;
            }
        case let stack as UIStackView:
            let orientationProp = diff.next.element.props["orientation"];
            if case .mixed(let prop) = orientationProp! {
                let orientation = prop.value.string!;
                stack.axis = orientation == "horizontal" ? .horizontal : .vertical;
            }
            
            for index in 0..<children.count {
                let child = children[index];
                stack.insertArrangedSubview(child.node, at: index);
            }
        default:
            return;
        }
    }
    
    func getNodes(_ commit: Commit) -> [RenderResult] {
        let node = nodes[commit.id]
        if (node != nil) {
            return [RenderResult(commit: commit, node: node!)];
        }
        
        return commit.children.flatMap { commit in
            return getNodes(commit);
        }
    }
    
    public func render(_ diff: Diff) -> [RenderResult] {
        let node: UIView? = nodes[diff.next.id] ?? createNode(diff);
        let children = diff.diffs.flatMap { diff in render(diff) }

        let hasDiff = diff.next.version != diff.prev.version;
      
        if (node == nil) {
            if (!hasDiff) {
                return getNodes(diff.next);
            }
            else {
                return children;
            }
        }

        updateNode(node: node, diff: diff, children: children);
        
        if (!diff.next.pruned) {
            return [RenderResult(commit: diff.next, node: node!)];
        }
        
        //removeNode(diff, node);
        return [];
    }
}
